### PR TITLE
WIP: Use the rnix "rowan" branch

### PR DESCRIPTION
**Pull request by jD91mZM2**
**Don't merge yet.**

I'm currently exploring a library called "rowan" in rnix, which supplies a builder for ASTs making the code prettier, but also makes comments completely separate tokens.
Sadly this library is currently slower because it's not using an arena-based approach, but that should hopefully be possible to do in rowan without modifying rnix or this project.
Eventually this will become the main version of rnix, so I fixed this project now :smile:

### CODE REVIEW

**Comment by jD91mZM2** [comment]
CI is broken because of the git dependency, I think you had a patch to carnix for that?

**Comment by tazjin** [comment]
Interesting, I have to take a more detailed look at what this actually implies, but I have two superficial comments:

1. Please keep the `#[macro_use]` annotation, `nixdoc` should build on NixOS stable which does not have the Rust version that drops this requirement.
2. Please rewrite the commit message into the [format I use](https://gist.github.com/stephenparish/9941e89d80e2bc58a153) (e.g. `refactor: Use rowan-based rnix dependency` with a body explaining the changes a little bit). I like keeping this uniform so people can see what is going on on a high-level without having to read through PRs.

> CI is broken because of the git dependency, I think you had a patch to carnix for that?

Yep, `carnix` keeps some of the query parameters appended by `cargo` in the git URL and `nix-prefetch-git` then fails to clone the repository.

Based on my [carnix devel branch](https://github.com/tazjin/carnix/tree/devel), the (kinda brute-force) patch is:

```
diff --git a/src/prefetch.rs b/src/prefetch.rs
index 0078cb4..dab20c2 100644
--- a/src/prefetch.rs
+++ b/src/prefetch.rs
@@ -182,6 +182,9 @@ impl Crate {
 
         println!("Prefetching {} ({})", self.name, cached_url);
         debug!("cached_url = {:?}", cached_url);
+
+        let url = url.split("?").next().unwrap();
         let prefetch = Command::new("nix-prefetch-git")
             .args(&[ "--url", url, "--rev", rev ])
             .output();
```

File: src/main.rs [hunk 1]
```diff
@@ -76,33 +82,55 @@ struct DocItem {
...
+        // Check if it's a comment
+        match node.kind() {
+            NodeType::Token(Token::Comment) => {
+                // Get the content and trim leading # or /*
+                let mut content = node.leaf_text().map(SmolStr::as_str).unwrap_or_default();
+                if content.starts_with('#') && allow_single_line {
+                    content = &content[1..];
+                } else if content.starts_with("/*") {
+                    content = &content[2..];
+                    if content.ends_with("*/") {
```
Comment (tazjin) [inline src/main.rs]: Can it ever not end with `*/`? Wouldn't that be a syntax error? :thinking:

Comment (jD91mZM2) [reply src/main.rs]: You're right, it can't. I'm too paranoid :stuck_out_tongue:

**Comment by asymmetric** [comment]
Closed by #36.

:heart:

**Pull request closed without merging**